src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs

```diff
@@ -207,6 +183,75 @@ int IVsEditorFactory.Close()
+            // next the new window's title and that [Read Only] is appended to title.
+            var readOnlyStatus = READONLYSTATUS.ROSTATUS_NotReadOnly;
+            if (ErrorHandler.Succeeded(textBuffer.GetStateFlags(out var textBufferFlags)) &&
+                0 != (textBufferFlags & ((uint)BUFFERSTATEFLAGS.BSF_FILESYS_READONLY | (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY)))
+            {
+                readOnlyStatus = READONLYSTATUS.ROSTATUS_ReadOnly;
+            }
+
+            return physicalView switch
+            {
+                FormPhysicalViewName => GetWinFormsDesignerData(vsHierarchy, itemId, textBuffer, readOnlyStatus, initializeView),
+                CodePhysicalViewName => GetCodeEditorData(textBuffer, readOnlyStatus, initializeView),
+                _ =>  (string.Empty, Guid.Empty, IntPtr.Zero),
+            };
+        }
+
+        private (string editorCaption, Guid commandUI, IntPtr documentView) GetWinFormsDesignerData(IVsHierarchy vsHierarchy, uint itemid, IVsTextBuffer textBuffer, READONLYSTATUS readOnlyStatus, bool initializeView)
+        {
+            var loaderName = GetWinFormsLoaderName(vsHierarchy);
+            var designerService = (IVSMDDesignerService)_oleServiceProvider.QueryService<SVSMDDesignerService>();
+            var designerLoader = (IVSMDDesignerLoader)designerService.CreateDesignerLoader(loaderName);
+            if (designerLoader is null)
+            {
+                return GetCodeEditorData(textBuffer, readOnlyStatus, initializeView);
+            }
+
+            try
+            {
+                if (initializeView)
+                {
+                    designerLoader.Initialize(_oleServiceProvider, vsHierarchy, (int)itemid, textBuffer as IVsTextLines);
+                }
+                var editorCaption = designerLoader.GetEditorCaption((int)readOnlyStatus);
+
+                var designer = designerService.CreateDesigner(_oleServiceProvider, designerLoader);
+
+                var documentView = initializeView ? Marshal.GetIUnknownForObject(designer.View) : IntPtr.Zero;
+                var commandUI = designer.CommandGuid;
+
+                return (editorCaption, commandUI, documentView);
+            }
+            catch
+            {
+                designerLoader.Dispose();
+                throw;
+            }
+        }
+
+        private (string editorCaption, Guid commandUI, IntPtr documentView) GetCodeEditorData(IVsTextBuffer textBuffer, READONLYSTATUS readOnlyStatus, bool initializeView)
+        {
+            var editorAdaptersFactoryService = _componentModel.GetService<IVsEditorAdaptersFactoryService>();
+            var codeWindow = editorAdaptersFactoryService.CreateVsCodeWindowAdapter(_oleServiceProvider);
+
+            if (initializeView)
+            {
+                codeWindow.SetBuffer((IVsTextLines)textBuffer);
+            }
+
+            codeWindow.GetEditorCaption(readOnlyStatus, out var editorCaption);
```

Comment:
Can we add an assert here that readOnlyStatus isn't READONLYSTATUS.ROSTATUS_Unknown? Because if we're calling this with that, then it's going to query the underlying buffer inside of GetEditorCaption; if we aren't setting that because initializeView = false, then we're not actually querying anything meaningful in the first place, right?